src/eir.rs

```diff
@@ -0,0 +1,250 @@
+use bytes::*;
+use bytes::buf::BufExt;
+use enumflags2::BitFlags;
+use num_derive::FromPrimitive;    
+use num_traits::FromPrimitive;
+use thiserror::Error;
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, BitFlags, Eq, PartialEq)]
+pub enum EIRFlags {
+    LELimitedDiscoverableMode  = 1 << 0,
+    LEGeneralDiscoverableMode = 1 << 1,
+    BREDRNotSupported = 1 << 2,
+    ControllerSimultaneousLEBREDR = 1 << 3,
+    HostSimultaneousLEBREDR = 1 << 4,
+}
+
+#[derive(Debug)]
+pub struct EIRName {
+    name: String,
+    complete: bool,
+}
+
+impl EIRName {
+    fn short_name(name: String) -> Self {
+        EIRName {
+            name: name,
+            complete: false,
+        }
+    }
+    fn complete_name(name: String) -> Self {
+        EIRName {
+            name: name,
+            complete: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ManufacturerSpecificData {
+    company_identifier_code: u16,
+    data: Bytes,
+}
+
+#[derive(Debug)]
+pub struct EIR {
+    flags: Option<BitFlags<EIRFlags>>,
+    uuid16: Vec<u16>,
+    uuid32: Vec<u32>,
+    uuid128: Vec<u128>,
+    name: Option<EIRName>,
+    tx_power_level: Vec<i8>,
+    uri: Vec<String>,
+    manufacturer_specific_data: Vec<ManufacturerSpecificData>,
+}
```

Comment:
Consider making this `EIR` type represent a single EIR tag, and then return a `Result<Vec<EIR>>` from `parse_eir`.

Then you could write `EIR` as a tagged enum:

```rust
#[derive(Debug)]
#[non_exahustive]
pub struct EIR {
    Flags(BitFlags<EIRFlags>),
    Uuid16(Vec<u16>),
    Uuid32(Vec<u32>),
    // ...
}
```

Reply:
Yes, I was wondering what the best API would be here. Enums and Vec (or maybe HashSet?) might make it a bit harder to match. But it feels clunky having a struct full of optionals. I'll try with Enums and see how it goes.

Reply:
An attempt at changing the return type to Vec: https://github.com/ruediger/bluez-rs/commit/cc91569e0cd35147a54f80a90d53beff697ffe36 Please take a look if you prefer this style.

Reply:
Looks good! Thank you, and please keep it up. One note: `EIR::Name` has more than one property, and it's not very obvious what the boolean is for, so maybe consider making the properties of that one named?

```rust
  Name { name: String, complete: bool },
```

Also, should the name be a `String` or a `CString`? I don't think the Bluetooth spec says device names have to be in UTF-8, so it might be good to give the end-user a chance to deal with this themselves.